src/Elasticsearch.Extensions.Logging/ElasticsearchLogger.cs

```diff
@@ -0,0 +1,359 @@
+		{
+			_categoryName = categoryName;
+			_dataProcessor = dataProcessor;
+		}
+
+		private Regex _w3CFormat = new Regex(@"^[abcdef]{2}-[\dabcdef]{32}-([\dabcdef]{16})-[\dabcdef]{2}$",
+			RegexOptions.Compiled);
+
+		internal ElasticsearchLoggerOptions Options { get; set; } = default!;
+
+		internal IExternalScopeProvider ScopeProvider { get; set; } = default!;
+
+		public IDisposable BeginScope<TState>(TState state) => ScopeProvider.Push(state);
+
+		public bool IsEnabled(LogLevel logLevel) => Options.IsEnabled;
+
+
+		public void Log<TState>(LogLevel logLevel, EventId eventId, TState state, Exception exception,
+			Func<TState, Exception, string> formatter
+		)
+		{
+			try
+			{
+				if (!IsEnabled(logLevel)) return;
+
+				if (formatter == null) throw new ArgumentNullException(nameof(formatter));
+
+				// TODO: Want to render state values (separate from message) to pass to log event, for semantic logging
+				// Maybe render to JSON in-process, then queue bytes for sending to index ??
+
+				var elasticsearchData =
+					BuildLogEvent(_categoryName, logLevel, eventId, state, exception, formatter);
+
+				_dataProcessor.EnqueueMessage(elasticsearchData);
+			}
+			catch (Exception ex)
+			{
+				Console.Error.WriteLine("ElasticsearchLogger exception: {0}", ex);
+			}
+		}
+
+		private static void AddException(Exception exception, LogEvent logEvent)
+		{
+			// Use the full string of the exception, which includes both the outer stack trace and any
+			// inner exceptions and their stack trace. It also includes any additional values that some
+			// exceptions have in ToString() that is not in Message.
+			var stackTrace = exception.ToString();
+			logEvent.Error = new Error { Type = exception.GetType().FullName, Message = exception.Message, StackTrace = stackTrace };
+		}
+
+		private void AddScopeValues(LogEvent logEvent)
+		{
+			var scopeProvider = ScopeProvider;
+			if (Options.IncludeScopes && scopeProvider != null)
+			{
+				scopeProvider.ForEachScope((scope, innerData) =>
+				{
+					if (logEvent.Labels == null) logEvent.Labels = new Dictionary<string, object>();
+
+					if (logEvent.Scopes == null) logEvent.Scopes = new List<string>();
+
+					var isFormattedLogValues = false;
+					if (scope is IEnumerable<KeyValuePair<string, object>> scopeValues)
+					{
+						foreach (var kvp in scopeValues)
+						{
+							if (kvp.Key == "{OriginalFormat}")
+							{
+								isFormattedLogValues = true;
+								continue;
+							}
+
+							if (CheckTracingValues(logEvent, kvp)) continue;
+
+							logEvent.Labels[kvp.Key] = FormatValue(kvp.Value);
+						}
+					}
+
+					var formattedScope = isFormattedLogValues ? scope.ToString() : FormatValue(scope);
+					logEvent.Scopes.Add(formattedScope);
+				}, logEvent);
+			}
+		}
+
+		private void AddStateValues<TState>(TState state, LogEvent logEvent)
+		{
+			if (state is IEnumerable<KeyValuePair<string, object>> stateValues)
```

Comment:
Is `state` always guaranteed to be an `IEnumerable<KeyValuePair<string, object>>`? Should something happen if it isn't e.g. log an error?

Reply:
State could be anything, called from the public Log<TState>() method.  However usually it is not called directly, and either the high performance LoggerMessage.Define<>() is used, or one of the convenience extension methods, e.g. LogWarning().

The usual calls convert the passed in parameters to an internal state class, which is what you get access to in a logger provider. If you are just logging the message you can use the passed in formatter function.

However, for semantic logging you want to get out the individual parameter values, which I then put into the ECS labels dictionary. To do this we assume if the class implements IEnumerable<KeyValuePair<string,object>>, which the internal Microsoft class does, then we can use it to get the semantic labels and values.

There is a similar assumption made about BeginScope<TState>() to pull out additional semantic values. As we don't have access to the actual internal code, this is the best we can do and just handle it gracefully if the internals ever change. 

Nic Blumhardt blogged about it for Seq handling of Scope: https://nblumhardt.com/2016/11/ilogger-beginscope/